### ISSUE TITLE
`schema_definition` metadata field is lost when event goes through disk buffers 

### ISSUE THREAD

**Issue by StephenWakely** [vectordotdev/vector#18574]
The `schema_definition` field in EventMetadata is not being persisted in the proto files and thus an event will lose the schema if it is stored in a disk buffer.

There will be challenges associated with fixing this.

- The `schema_definition` takes up a fairly significant amount of memory. It is stored in an `Arc` since many events will share the same schema. If we persist the schema directly in the buffer, the buffers will fill up very quickly. 
- A possible solution could be to store the schema in memory and merely persist a reference in the buffer. Care need to be taken to ensure the reference remains valid should Vector be updated and restarted. 
- Should the schema be stored in a separate buffer and the reference point to this buffered definition?

**Comment by tobz** [comment]
One question I'd have is: are these schema definitions tied to the configuration used to generate an event, or are they always fixed relative to the Vector process? (i.e. always the same for an event generated by source X for version Y of Vector)

**Comment by StephenWakely** [comment]
The schema definition is for the most part tied to the configuration used, but for each event it would depend on the route the event has taken through the topology. Eg, an event that has come `stdin -> route -> remap -> datadog_logs` would have a different schema to `syslog -> route -> remap -> datadog_logs`.

We do need to take into account the version of Vector. An event created by `v0.31` could well have a different schema definition to one created by `v0.32` if the latter version updates the schema definition for the source.

Thinking about it, I'm pretty sure the only part of the the definitions that are required at runtime is the list of meanings. A sink may refer to a field `get_by_meaning("timestamp")` and this needs to be looked up to determine where the actual field is in the event. This may not be too big, at most a map of perhaps 3 or 4 strings for the meaning to the path in the event. @fuchsnj can you confirm?

**Comment by fuchsnj** [comment]
> @fuchsnj can you confirm?

Yes, I believe we would only need the semantic meanings (a map from meanings to paths). We would need to solve some edge cases where the required meanings changed, but the events currently in the disk buffer doesn't contain new ones yet. I think this could result in a panic on sources when they try to fetch a required meaning.

**Comment by a-abella** [comment]
Got pretty burned by this this weekend. Debug logging wasn't much help either, all I was seeing was 1-byte POST payloads to my endpoints. Took a lot of trial and error before realizing that disabling disk buffers solved the problem.

I see there was a PR to document the namespacing+disk buffer deficiency #21813 -- it looks like that doc change got blown away by a subsequent PR #22408

**Comment by pront** [comment]
> I see there was a PR to document the namespacing+disk buffer deficiency [#21813](https://github.com/vectordotdev/vector/pull/21813) -- it looks like that doc change got blown away by a subsequent PR [#22408](https://github.com/vectordotdev/vector/pull/22408)

Thank you for bringing this to our attention. We will add the warning back. I also realized that `schema` options are not documented at all.
